ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/StatementExecutor.java

```diff
@@ -444,4 +491,89 @@ private void terminateQuery(final TerminateQuery terminateQuery) throws Exceptio
         new CommandStatus(CommandStatus.Status.TERMINATED, "Query terminated")
     );
   }
+
+  private void terminateCluster(
+      final List<String> keepSources,
+      final List<String> deleteSources
+  ) {
+    // Terminate all queries
+    getListForSet(ksqlEngine.getAllLiveQueries()).forEach(
+        queryMetadata -> {
+          if (queryMetadata instanceof PersistentQueryMetadata) {
+            final PersistentQueryMetadata persistentQueryMetadata
+                = (PersistentQueryMetadata) queryMetadata;
+            ksqlEngine.terminateQuery(persistentQueryMetadata.getQueryId(), true);
+          }  else {
+            queryMetadata.close();
+          }
+        }
+    );
+
+    // if we have the explicit list of stream/table to delete.
+    final MetaStore metaStore = ksqlEngine.getMetaStore();
+    if (!deleteSources.isEmpty()) {
+      deleteSources
+          .forEach(sourceName -> deleteSource(sourceName, metaStore.getSource(sourceName)));
+    } else if (!keepSources.isEmpty()) {
+      metaStore.getAllStructuredDataSources().forEach(
+          (sourceName, structuredDataSource) -> {
+            if (!keepSources.contains(sourceName)) {
+              deleteSource(sourceName, structuredDataSource);
+            }
+          }
+      );
+    } else {
+      metaStore.getAllStructuredDataSources().forEach(this::deleteSource);
+    }
+  }
+
+  // This is needed because the checkstyle complains if we create this in place.
+  List<QueryMetadata> getListForSet(final Set<QueryMetadata> queryMetadataSet) {
+    final List<QueryMetadata> queryMetadataList = new ArrayList<>();
+    queryMetadataList.addAll(ksqlEngine.getAllLiveQueries());
+    return queryMetadataList;
+  }
+
+  private void deleteSource(
+      final String sourceName,
+      final StructuredDataSource structuredDataSource) {
+    final DdlCommand ddlCommand;
+    if (structuredDataSource.getDataSourceType() == DataSourceType.KSTREAM) {
+      ddlCommand = new DropSourceCommand(
+          new DropStream(QualifiedName.of(sourceName), false, true),
```

Comment:
you could extract a method to get the `DropStream` or `DropTable`. Put the if block in there. Then this code becomes much simpler and more condense as the rest of it is identical

Reply:
Moved it into a method as you suggested.